metricbeat/module/munin/munin.go

```diff
@@ -0,0 +1,109 @@
+package munin
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+const (
+	unknownValue = "U"
+)
+
+// Node connection
+type Node struct {
+	conn net.Conn
+
+	writer io.Writer
+	reader *bufio.Reader
+}
+
+// Connect with a munin node
+func Connect(address string, timeout time.Duration) (*Node, error) {
+	conn, err := net.DialTimeout("tcp", address, timeout)
+	if err != nil {
+		return nil, err
+	}
+	n := &Node{conn: conn,
+		writer: conn,
+		reader: bufio.NewReader(conn),
+	}
+	// Cosume and ignore first line returned by munin, it is a comment
+	// about the node
+	bufio.NewScanner(n.reader).Scan()
+	return n, err
+}
+
+// Close node connection relasing its resources
+func (n *Node) Close() error {
+	return n.conn.Close()
+}
+
+// List of items exposed by the node
+func (n *Node) List() ([]string, error) {
+	_, err := io.WriteString(n.writer, "list\n")
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(n.reader)
+	scanner.Scan()
+	return strings.Fields(scanner.Text()), scanner.Err()
+}
+
+// Fetch metrics from munin node
+func (n *Node) Fetch(items ...string) (common.MapStr, error) {
+	var errs multierror.Errors
+	event := common.MapStr{}
+
+	for _, item := range items {
+		_, err := io.WriteString(n.writer, "fetch "+item+"\n")
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		scanner := bufio.NewScanner(n.reader)
+		scanner.Split(bufio.ScanWords)
+		for scanner.Scan() {
+			name := strings.TrimSpace(scanner.Text())
+
+			// Munin delimites metrics with a dot
+			if name == "." {
+				break
+			}
+
+			name = strings.TrimSuffix(name, ".value")
+
+			scanner.Scan()
+			value := scanner.Text()
+
+			key := fmt.Sprintf("%s.%s", item, name)
+
+			if value == unknownValue {
+				errs = append(errs, errors.Errorf("unknown value for %s", key))
+				continue
+			}
+			if f, err := strconv.ParseFloat(value, 64); err == nil {
+				event.Put(key, f)
```

Comment:
for the record, this is a candidate for `safemapstr.Put`: https://github.com/elastic/beats/pull/6434 as soon as it's in

Reply:
FYI, the PR went in already, so you can update these :)

Reply:
Good point, but I think dots are not allowed in munin metric names.
We could use `safemapstr` in any case to protect metricbeat against incorrect implementations of munin nodes of plugins, wdyt?

Reply:
I'm wondering if the float parsing will make a difference in ES. If we have a value 3 and we parse it it will still put 3 into the event and on the ES side it will assume it's an integer so conflicts will happen afterwards. Sounds like a good use case for the local fields definitions.

Reply:
Umm, so what do you recommend to do here?

Reply:
I've checked metrics format, sounds like we don't need `safemapstr.Put` here

Reply:
I think what I'm getting at we could even skip this code part here and it would make no difference? If someone wants to make sure the mapping is correct he has to use https://github.com/elastic/beats/pull/6024

Please double check the above statement as I don't know much about the munin output.